### PR TITLE
Pass through to facts

### PR DESCRIPTION
**Pull request by lexidor**
This PR adds an optional configuration flag to make hhvm-autoload emit a shim file which asks Facts for the autoload map. I have explained the rationale in the README, but in short.

Hhvm-autoload has become a defacto pseudo whole code base reflection mechanism. The autoload map is generally useful, even when using watchman autoloading. However, when using a native autoloader, developers don't tend to regen the map. This results in anything from skipping tests in Hacktest to hhast linters to failing to find a class. We can't simply update projects to use Facts instead, since Facts is not always enabled. Having hhvm-autoload asking Facts for users using native autoloading seems like a good compromise.

I was unsure about whether to process the output from Facts. I do not want to break BC, unless there is a good reason for it. Preserving BC would entail performing the filtering on each request, rather than when the map is being emitted. Depending on the size of the code base, this might be an unacceptable performance regression. The implementation complexity would also be quite large.

Hhvm-autoload can be configured via the top-level hh_autoload.json and implicitly through the config files of vendor dependencies. Facts ignores these config files and honors the .hhvmconfig.hdf file. We could make a best effort to codegen in all the rules of all the hh_autoload.json configs when emitting the autoload.hack file. This would be a lot of unpleasant to read code, since we can't rely on hack-codegen. Alternatively, we could try to load the hh_autoload.json files at runtime and filter with more general code. This would be easier on the eyes. However, this runs into trouble when repo auth mode is enabled and the source directory is not on disk. Again, we could theoretically dump the json decoded information of these files in a constant in autoload.hack to work around that. This feels like a lot of work to "protect" users from seeing classes they shouldn't see if their native autoload query is wider than their hhvm-autoload configuration. And even if implemented, we still can't protect against the inverse, the autoload query being narrower than the hhvm-autoload configuration. Let's be clear to the users of this library:
**You either need to write a very specific native autoload query or accept that the map contains things it didn't contain before.**

Maybe in future, add a function which returns a non-lowercased map using the canonical class, function, and type names.

### CODE REVIEW

**Comment by lexidor** [comment]
Oh, I based this branch of #71 instead of master. I don't think #71 needs changes, so this mistake shouldn't matter much.

File: README.md [hunk 1]
```diff
@@ -5,10 +5,11 @@ The autoloader for autoloading classes, enums, functions, typedefs, and constant
 Usage
 =====
 
-1. Add an `hh_autoload.json` file (see section below) and optionally remove your configuration from composer.json
+1. Add an `hh_autoload.json` file (see section below)
 2. `composer require hhvm/hhvm-autoload`
-3. Replace any references to `vendor/autoload.php` with  `vendor/autoload.hack` and call `Facebook\AutoloadMap\initialize()`
-4. To re-generate the map, run `vendor/bin/hh-autoload`, `composer dump-autoload`, or any other command that generates the map
+3. Require the autoload file from your entrypoint functions using `require_once __DIR__ . '/..(repeat as needed)/vendor/autoload.hack`
```
Comment (fredemmott) [inline README.md:10]: while you're here, parens :)

File: README.md [hunk 2]
```diff
@@ -32,9 +33,10 @@ The following settings are optional:
...
- - `"failureHandler:" classname<Facebook\AutoloadMap\FailureHandler>` - use the specified class to handle definitions that aren't the Map. Your handler will not be invoked for functions or constants
-   that aren't in the autoload map and have the same name as a definition in the global namespace. Defaults to none.
+ - `"failureHandler:" classname<Facebook\AutoloadMap\FailureHandler>` - use the specified class to handle definitions that aren't the Map. Defaults to none.
  - `"devFailureHandler": classname<Facebook\AutoloadMap\FailureHandler>` - use a different handler for development environments. Defaults to the same value as `failureHandler`.
+ - `"parser:" any of [ext-factparse]"` - select a parser to use, but there is only one valid option. Defaults to a sensible parser.
+ - `"useFactsIfAvailableAndDoNotEmitAStaticMap": false` - use ext-facts (HH\Facts\...) to back Facebook\AutoloadMap\Generated\map() instead of a codegenned dict. See _Use with HH\Facts_ for more information about this mode.
```
Comment (fredemmott) [inline README.md:39]: just `useFactsIfAvailable` ?

File: README.md [hunk 3]
```diff
@@ -90,12 +92,24 @@ Information you may need is available from:
  - `Facebook\AutoloadMap\Generated\root()`: the directory containing the
     project root, i.e. the parent directory of `vendor/`
 
+Use with HH\Facts
```
Comment (fredemmott) [inline README.md:95]: The \ probably needs escaping

**Review by fredemmott — COMMENTED** [review]
- DO_NOT_REQUIRE_THIS_FILE.hack should probably be deleted in parent diff
- if actually using the native autoloader, IMO it's fine to completely ignore any config in `hh_autoload.json`. We want HHVM's view to be authoritative. If not using facts, HHVM's view == whatever's in hh_autoload, and if using facts, it's not.
- there's potentially another option: build the map, but support facts as a fallback handler in a similar way to how `hh_client` can be used as a fallback handler. Does that work for you? Perhaps supporting both would be good

**Comment by lexidor** [comment]
> there's potentially another option: build the map,
> but support facts as a fallback handler in a similar way to how `hh_client`
>can be used as a fallback handler. Does that work for you?
> Perhaps supporting both would be good

Facts being a fallback handler is a great idea. I'll see if I can make it happen.
About building the map, the reason for not emitting the map was that `Generated\map()` should never be out of date. Asking Facts for it live guarantees that anyone using it for reflection purposes gets a complete answer.

**Comment by lexidor** [comment]
When you use the Facts autoloader, you don't use `Generated\map()` for actual autoloading. `AutoloadMap\initialize()` becomes a noop.

**Comment by fredemmott** [comment]
AIUI we essentially do something something very similar: use ext_facts in dev, built in prod, with a system called 'arc facts' abstracting over them.

@jthemphill any thoughts on this PR? LGTM

**Comment by lexidor** [comment]
Before merging, HH\Facts has no hhi. Should I wait for that hhi or suppress the unbound name errors?

File: README.md (hunk 1 above)
Comment (lexidor) [reply README.md]: Added parens both for the require_once statement and grouping the `/..`. I didn't know which one you meant.

**Comment by jthemphill** [comment]
Oof, give me a sec to move the hhi into open source.

**Comment by lexidor** [comment]
> Oof, give me a sec to move the hhi into open source.

While you're there, close the issue in facebook/hhvm for it. It should be the latest issue.

**Comment by lexidor** [comment]
My plan is as follows. Wait for the hhi commit. Snip the hhi into my local checkout to see if it would typecheck. The add unbound symbol suppression comments. Then we don't have to upgrade out hhvm version to latest to get this version of hhvm-autoload. Minimum version can be 4.109 (the release version of Facts).

**Comment by lexidor** [comment]
I have typechecker with the hhi from https://github.com/facebook/hhvm/pull/8853/files

**Pull request merged**